File: 04/04.py
from pathlib import Path
from typing import List
from functools import reduce
import logging

logging.basicConfig()
log = logging.getLogger(__name__)
log.setLevel(level=logging.WARNING)


def first_question(input_data: str) -> int:
    lines = input_data.splitlines()
    contained_ranges = 0
    for l in lines:
        first_range, second_range = l.split(",")
        first_range = list(map(lambda x: int(x), first_range.split("-")))
        second_range = list(map(lambda x: int(x), second_range.split("-")))
        if (
            first_range[0] >= second_range[0]
            and first_range[1] <= second_range[1]
            or first_range[0] <= second_range[0]
            and first_range[1] >= second_range[1]
        ):
            contained_ranges += 1
        log.debug(f"first: {first_range} second:{second_range}")
    return contained_ranges


def second_question(input_data: str) -> int:
    lines = input_data.splitlines()
    overlaps = 0
    for l in lines:
        first_range, second_range = l.split(",")
        first_range = list(map(lambda x: int(x), first_range.split("-")))
        second_range = list(map(lambda x: int(x), second_range.split("-")))
        if (
            first_range[0] <= second_range[0]
            and first_range[1] <= second_range[1]
            and first_range[1] >= second_range[0]
            or first_range[0] >= second_range[0]
            and first_range[0] <= second_range[1]
            and first_range[1] >= second_range[1]
            or first_range[0] >= second_range[0]
            and first_range[1] <= second_range[1]
            or first_range[0] <= second_range[0]
            and first_range[1] >= second_range[1]
        ):
            overlaps += 1
            log.debug("OVERLAP")
        log.debug(f"first: {first_range} second:{second_range}")
    return overlaps


if __name__ == "__main__":

    input_data = (Path(__file__).parent / "INPUT").read_text("UTF8")
    print(first_question(input_data))
    print(second_question(input_data))

File: 04/04.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let mut lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer1) = first_question(lines) {
        println!("{}", answer1);
    };
    lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer2) = second_question(lines) {
        println!("{}", answer2);
    };
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn first_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<u32> {
    let mut contained_ranges: u32 = 0;
    for line in lines {
        let l = line?;
        let ranges: Vec<Vec<u32>> = l
            .split(",")
            .map(|range| {
                range
                    .split("-")
                    .map(|value| value.parse::<u32>().unwrap())
                    .collect()
            })
            .collect();
        let first_range = &ranges[0];
        let second_range = &ranges[1];
        if first_range[0] >= second_range[0] && first_range[1] <= second_range[1]
            || first_range[0] <= second_range[0] && first_range[1] >= second_range[1]
        {
            contained_ranges += 1;
        }
    }

    return Ok(contained_ranges);
}

fn second_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<u32> {
    let mut overlaps: u32 = 0;
    for line in lines {
        let l = line?;
        let ranges: Vec<Vec<u32>> = l
            .split(",")
            .map(|range| {
                range
                    .split("-")
                    .map(|value| value.parse::<u32>().unwrap())
                    .collect()
            })
            .collect();
        let first_range = &ranges[0];
        let second_range = &ranges[1];
        if first_range[0] <= second_range[0]
            && first_range[1] <= second_range[1]
            && first_range[1] >= second_range[0]
            || first_range[0] >= second_range[0]
                && first_range[0] <= second_range[1]
                && first_range[1] >= second_range[1]
            || first_range[0] >= second_range[0] && first_range[1] <= second_range[1]
            || first_range[0] <= second_range[0] && first_range[1] >= second_range[1]
        {
            overlaps += 1;
        }
    }
    return Ok(overlaps);
}
